src/Api.php

```diff
@@ -145,25 +166,179 @@ protected function computeHmac($hmac, $fields)
     {
-        $result = array();
+        $result = [];
         foreach ($array as $key => $value) {
             $result[] = sprintf('%s=%s', $key, $value);
         }
+
         return implode('&', $result);
     }
 
+    public function getOptions(): array
+    {
+        return $this->options;
+    }
+
     /**
-     * @return array
+     * @See documentation ManuelIntegrationVerifone_PayboxSystem_V8 5.3.4 Values checking
+     *
+     * @param GetHttpRequest $httpRequest
+     *
+     * @return bool
      */
-    public function getOptions()
+    public function verify($httpRequest)
     {
-        return $this->options;
+        $context = [
+            'httpRequestUri' => $httpRequest->uri,
+            'httpRequestQuery' => $httpRequest->query,
+        ];
+        if (false === $this->doesTheResponseReallyComeFromPaybox($httpRequest)) {
+            $message = 'Notification invalid: the response doesn\'t come from Paybox.';
+            $this->logger->error($message, $context);
+            throw new HttpResponse($message);
+        }
+
+        return true;
+    }
+
+    /**
+     * @param GetHttpRequest $httpRequest
+     *
+     * @return bool
+     */
+    private function doesTheResponseReallyComeFromPaybox($httpRequest)
+    {
+        // checks paybox signature
+        $this->logger->notice('checkPayboxSignature', [
+            'return' => $this->checkPayboxSignature($httpRequest),
+        ]);
+        if (false === $this->checkPayboxSignature($httpRequest)) {
+            return false;
+        }
+        // checks origin ip address
+        $this->logger->notice('checkOriginIpAddress', [
+            'return' => $this->checkOriginIpAddress($httpRequest),
+        ]);
+        if (false === $this->checkOriginIpAddress($httpRequest)) {
+            return false;
+        }
+
+        return true;
     }
 
+    /**
+     * @return bool
+     */
+    private function checkOriginIpAddress(GetHttpRequest $httpRequest)
+    {
+        return \in_array($httpRequest->clientIp, self::PAYBOX_IP_ADDRESSES, true);
+    }
+
+    /**
+     * @return bool
+     */
+    private function checkPayboxSignature(GetHttpRequest $httpRequest)
+    {
+        if (false === isset($httpRequest->query['signature'])) {
+            $this->logger->critical('checkPayboxSignature: query parameter \'signature\' not set', [
+                'query' => $httpRequest->query,
+            ]);
+
+            return false;
+        }
+
+        // init signature
+        $signature = $httpRequest->query['signature'];
+        $signatureLength = \strlen($signature);
+        if ($signatureLength > 172) {
+            $signatureInitialized = base64_decode(urldecode($signature));
+        } elseif (172 == $signatureLength) {
+            $signatureInitialized = base64_decode($signature);
+        } elseif (128 == $signatureLength) {
+            $signatureInitialized = $signature;
+        } else {
+            $signatureInitialized = null;
```

Comment:
`signatureInitialized` is set to null but a few line later we check for false, what is the expected type of `signatureIntialized` ? `string|bool|null` ?

Reply:
You're right, I fixed with :
```
        } else {
                $this->logger->error('Fail to base_decode signature', [
                    'signature' => $signature,
                ]);

                return false;
        }

```